### DISCUSSION TITLE
How do I make votes table for questions?

### DISCUSSION THREAD

**Discussion by Domino-98** [supabase/supabase#6372] [Questions] [answered]
I was thinking about using onConflict to combine 2 columns but it's not possible. I need to have 2 unique columns: question_id and user_id so each user can only vote once. If values for them already exist, update vote value column, otherwise add new row. How can I edit this table so it would be possible?
![Przechwytywanie](https://user-images.githubusercontent.com/64014834/162485522-a2b726e1-8d61-4be4-9711-ecf34b11635f.JPG)

**Comment by GaryAustin1** [comment, marked as answer]
You can create a unique index on two columns...
https://github.com/supabase/supabase/discussions/4546#discussioncomment-1842024
https://www.postgresqltutorial.com/postgresql-indexes/postgresql-unique-index/  (about 2/3 down)

**Reply by Domino-98** [reply]
Thanks. It works almost as expected but the problem is that it throws an error with onConflict instead of updating row so I have to add another query for updating row in error bracket.

**Reply by GaryAustin1** [reply]
Do you have upsert flag on?   Can you show your call.

**Reply by Domino-98** [reply]
> Do you have upsert flag on? Can you show your call.

example from my code:
`
const { error, data } = await supabase.from("votes_question").upsert(
            {
              question_id: question.id,
              user_id: user.id,
              unique_question_user: null,
              upvoted: true,
              downvoted: false,
            },
            { onConflict: "unique_question_user" }
          );
          if (error) {
            const { error, data } = await supabase
              .from("votes_question")
              .update({
                upvoted: true,
                downvoted: false,
              })
              .match({ question_id: question.id, user_id: user.id });
`

query I ran on this table: CREATE UNIQUE INDEX unique_question_user on votes_question (question_id, user_id)

**Reply by GaryAustin1** [reply]
I linked the wrong example in the 2nd article.   That was a unique index on two columns...
```CONSTRAINT question_user_unique UNIQUE (question_id, user_id)``` is really what you need for the conflict to work, versus just get an error for violating index....
https://www.techonthenet.com/postgresql/unique.php

**Reply by Domino-98** [reply]
> I linked the wrong example in the 2nd article. That was a unique index on two columns... `CONSTRAINT question_user_unique UNIQUE (question_id, user_id)` is really what you need for the conflict to work, versus just get an error for violating index.... https://www.techonthenet.com/postgresql/unique.php

It still doesn't work for me.
query I ran: `ALTER TABLE votes_question ADD CONSTRAINT question_user_unique UNIQUE (question_id, user_id)`
Do I have to create 'question_user_unique' column with unique id? I tried it and it doesn't work. 409 onconflict error. It only inserts row and I can't change vote value.

**Reply by GaryAustin1** [reply]
This worked for me (note onConflict has both columns listed):
``` .upsert({num1:8,num2:12,num3:6},{onConflict:'num1,num2'})```
Where I added a constraint: ```alter table test8 add constraint num1_num2 unique (num1,num2)```
You may need to delete the index if still there also.

It would add new row to table if one, or none of the columns is the same as an existing, but would update if both are the same as an existing row.

**Reply by Domino-98** [reply]
Thanks! I didn't know I can pass 2 columns there.

**Reply by GaryAustin1** [reply]
You still have to add the constraint on unique (x,y) though or it would just evaluate each column separately if they were individually defined unique.

**Reply by Domino-98** [reply]
I did it with `ALTER TABLE votes_question ADD CONSTRAINT question_user_unique UNIQUE (question_id, user_id)` and then
`
          const { error, data } = await supabase.from("votes_question").upsert(
            {
              question_id: question.id,
              user_id: user.id,
              upvoted: true,
              downvoted: false,
            },
            { onConflict: "question_id,user_id" }
          );
`
It seems to be working now.
